draft-ietf-jsonpath-base.md

```diff
@@ -502,6 +502,14 @@ to the JSONPath processing (e.g., index values and steps) MUST be
 within the range of exact values defined in I-JSON {{-i-json}}, namely
 within the interval \[-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1].
 
+2. Any function extensions used by the query MUST be supported by the JSONPath
+implementation, MUST have the correct number of arguments of the correct kinds
```

Comment:
I think this is reversed. It sounds like if `foo()` is in the query, then the implementation is required to support it.  But an implementation can't support functions it doesn't know about.

What I think you want is that queries must only use functions that are supported by the implementation.  This way the requirement is on the query author, not the implementation.

Reply:
The query author cannot know what the implementation will be that interprets the query.

The requirement does need to be on the implementation: Detect an unsupported function and handle it as a (static -- before the query is even performed) error.

Reply:
Yes, that's was what I intended.

Reply:
@glyn, which is what you intended?

As an implementor, I can't know what authors are doing to use in their paths.  All I can do is offer support for specific functions.

The way the text is worded, it sounds like if my implementation doesn't support a function, my implementation is at fault, when it's really the author (who very likely _does_ have some sort of tooling documentation as to which functions are supported, even if they don't know the underlying implementation) who knowingly used an unsupported function.

I don't need users opening bugs saying that I MUST support some function because the spec requires I support any function they decide to put in their queries.

---

What I _think_ this is trying to say is that query authors need to only used functions that are explicitly supported by the tooling they're using.

Reply:
> Any function extensions used by the query MUST be supported by the JSONPath implementation.

This text can be read two ways:

1. The query MUST only use functions extensions that are supported.
2. The implementation MUST support any function extensions that may appear in the query.

What I think you mean is option 1.

Option 2 is a backwards requirement and cannot be fulfilled by any implementation.

Regardless, this text is not clear.

Reply:
> Detect an unsupported function and handle it as a (static -- before the query is even performed) error.

I don't disagree with you on this, but this text isn't about erroring on unknown functions.

Reply:
I get it now. Thanks @gregsdennis. I've started to tease that out with some more text.